### DISCUSSION TITLE
Infer query output based on the hook

### DISCUSSION THREAD

**Discussion by larskarbo** [trpc/trpc#3225] [Q&A]
Is there any way to take the hook (example `trpc.tasks.getAll`), and Infer the query output with TypeScript?

In TRPC v9 I could do this by using the string (`"tasks.getAll"`) and use the [type inference functions](https://trpc.io/docs/v9/infer-types) but it seems to be more tricky when it's not a string, but a nested reference.

I need this to create a custom hook abstraction above trpc.

**Comment by JonParton** [comment]
Hey @larskarbo!

You can still do this in V10 very similarly but instead of using a `.` separated string like in V9 just [chain indexes off the type](https://www.typescriptlang.org/docs/handbook/2/indexed-access-types.html). 

see the [Inference documentation post.create output example](https://trpc.io/docs/infer-types#:~:text=type%20PostCreateOutput%20%3D%20RouterOutput%5B%27post%27%5D%5B%27create%27%5D%3B)

in short

```ts
// Helper exported from tRPC
import type { inferRouterInputs, inferRouterOutputs } from '@trpc/server';

// Your exported router type
import type { AppRouter } from './server';

// Get the inferred types ... normally done in a central file and rexported to prevent 2 imports at the top! 
type RouterOutput = inferRouterOutputs<AppRouter>;

// Index the inferred type!
type TasksGetAllOutput = RouterOutput['tasks']['getAll']; // Infers output of your desired "tasks.getAll" route. 

```

**Reply by larskarbo** [reply]
Thank you! but I'm trying to infer the type without writing it twice. I need both the hook `trpc.tasks.getAll` _and_ the output type.

I'll give a bit more context to my situation.

In TRPC v9 I was able to create an abstracted hook `useInfiniteFetch` that worked like this:

```ts
const {
  allItems,
  isLoading,
  ref,
} = useInfiniteFetch([
  "admin.bounty.getBountyStepCompletions",
  {
    stepId: step.id,
  },
])
```

This hook had the exact same _input_ parameters as the normal useQuery hook, but it did some magic under the hood and returned a modified _output_ object (with `ref` and `allItems`)

It worked like this:

```ts
export const useInfiniteFetch = <T extends InfiniteQueryNames,>(
  ...props: Parameters<typeof trpc.useInfiniteQuery<T>>
): ReturnType<typeof trpc.useInfiniteQuery<T>> & {
  ref: ReturnType<typeof useInView>["ref"]
  allItems: InferQueryOutput<T>["items"] | undefined
} => {
  const query = trpc.useInfiniteQuery<T>(...props)

  const { ref } = useInView()

  // some magic here: flattening pages into allItems, fetching next page on scroll, etc...

  return {
    ...query,
    allItems,
    ref,
  }
}
```

This was a very useful hook, that let us abstract away code when adding infinite fetch queries.

I haven't been able to make this work for TRPC v10.
